PR: https://github.com/odoo/odoo/pull/24738

From: 10f1a1a0c451a3538ac77c6c06cd26fb5898fa9c
From: Xavier Morel
Date: 2019-05-27 12:00:39

Structural Changes: 3
Total Changes: 76

[REM] multi from ir.actions.act_window

Task 1843603

* src_model is redundant with binding_model_id
* multi -> binding_view_types (if empty => all views) (maybe should be
empty by default yo?)
* in convert, type => rec.get(type) but no @type possible on <act_window>...
* removed deprecated auto_refresh & auto_search (not used anywhere (?))

closes odoo/odoo#24738

Signed-off-by: Xavier Morel (xmo) <xmo@odoo.com>

================================= pseudo patch: =================================

--- a/odoo/addons/base/models/ir_actions.py
+++ b/odoo/addons/base/models/ir_actions.py
@@ -45,9 +45,9 @@ class IrActions(models.Model):
     binding_model_id = fields.Many2one('ir.model', ondelete='cascade',
                                        help="Setting a value makes this action available in the sidebar for the given model.")
     binding_type = fields.Selection([('action', 'Action'),
-                                     ('action_form_only', "Form-only"),
                                      ('report', 'Report')],
                                     required=True, default='action')
+    binding_view_types = fields.Char(default='list,form')
 
     def _compute_xml_id(self):
         res = self.get_external_id()
@@ -134,13 +134,13 @@ class IrActionsActWindow(models.Model):
     _sequence = 'ir_actions_id_seq'
     _order = 'name'
 
-    @api.constrains('res_model', 'src_model')
+    @api.constrains('res_model', 'binding_model_id')
     def _check_model(self):
         for action in self:
             if action.res_model not in self.env:
                 raise ValidationError(_('Invalid model name %r in action definition.') % action.res_model)
-            if action.src_model and action.src_model not in self.env:
-                raise ValidationError(_('Invalid model name %r in action definition.') % action.src_model)
+            if action.binding_model_id and action.binding_model_id.model not in self.env:
+                raise ValidationError(_('Invalid model name %r in action definition.') % action.binding_model_id.model)
 
     @api.depends('view_ids.view_mode', 'view_mode', 'view_id.type')
     def _compute_views(self):
@@ -180,8 +180,6 @@ class IrActionsActWindow(models.Model):
     res_id = fields.Integer(string='Record ID', help="Database ID of record to open in form view, when ``view_mode`` is set to 'form' only")
     res_model = fields.Char(string='Destination Model', required=True,
                             help="Model name of the object to open in the view window")
-    src_model = fields.Char(string='Source Model',
-                            help="Optional model name of the objects on which this action should be visible")
     target = fields.Selection([('current', 'Current Window'), ('new', 'New Window'), ('inline', 'Inline Edit'), ('fullscreen', 'Full Screen'), ('main', 'Main action of Current Window')], default="current", string='Target Window')
     view_mode = fields.Char(required=True, default='tree,form',
                             help="Comma-separated list of allowed view modes, such as 'form', 'tree', 'calendar', etc. (Default: tree,form)")
@@ -199,9 +197,7 @@ class IrActionsActWindow(models.Model):
                                  'act_id', 'gid', string='Groups')
     search_view_id = fields.Many2one('ir.ui.view', string='Search View Ref.')
     filter = fields.Boolean()
-    auto_search = fields.Boolean(default=True)
     search_view = fields.Text(compute='_compute_search_view')
-    multi = fields.Boolean(string='Restrict to lists', help="If checked and the action is bound to a model, it will only appear in the More menu on list views")
 
     @api.multi
     def read(self, fields=None, load='_classic_read'):

--- a/odoo/addons/base/tests/test_ir_actions.py
+++ b/odoo/addons/base/tests/test_ir_actions.py
@@ -253,54 +253,6 @@ class TestServerActions(TestServerActionsBase):
         self.assertEqual(self.test_country.vat_label, 'VatFromTest', 'vat label should be changed to VatFromTest')
 
 
-class TestActionBindings(common.TransactionCase):
-
-    def test_bindings(self):
-        """ check the action bindings on models """
-        Actions = self.env['ir.actions.actions']
-
-        # first make sure there is no bound action
-        self.env.ref('base.action_partner_merge').unlink()
-        bindings = Actions.get_bindings('res.partner')
-        self.assertFalse(bindings['action'])
-        self.assertFalse(bindings['report'])
-
-        # create action bindings, and check the returned bindings
-        action1 = self.env.ref('base.action_attachment')
-        action2 = self.env.ref('base.ir_default_menu_action')
-        action3 = self.env['ir.actions.report'].search([('groups_id', '=', False)], limit=1)
-        action1.binding_model_id = action2.binding_model_id \
-                                 = action3.binding_model_id \
-                                 = self.env['ir.model']._get('res.partner')
-
-        bindings = Actions.get_bindings('res.partner')
-        self.assertItemsEqual(
-            bindings['action'],
-            (action1 + action2).read(),
-            "Wrong action bindings",
-        )
-        self.assertItemsEqual(
-            bindings['report'],
-            action3.read(),
-            "Wrong action bindings",
-        )
-
-        # add a group on an action, and check that it is not returned
-        group = self.env.ref('base.group_user')
-        action2.groups_id += group
-        self.env.user.groups_id -= group
-
-        bindings = Actions.get_bindings('res.partner')
-        self.assertItemsEqual(
-            bindings['action'],
-            action1.read(),
-            "Wrong action bindings",
-        )
-        self.assertItemsEqual(
-            bindings['report'],
-            action3.read(),
-            "Wrong action bindings",
-        )
 
 
 class TestCustomFields(common.TransactionCase):

--- a/odoo/addons/base/views/ir_actions_views.xml
+++ b/odoo/addons/base/views/ir_actions_views.xml
@@ -86,7 +86,6 @@
                             </page>
                             <page name='advanced' string="Advanced Properties">
                                 <group>
-                                    <field name="multi"/>
                                     <field name="attachment_use"/>
                                     <field name="attachment"/>
                                 </group>
@@ -163,7 +162,6 @@
                             <field name="name"/>
                             <field name="xml_id" string="External ID"/>
                             <field name="res_model" string="Object"/>
-                            <field name="src_model" string="Source Object"/>
                         </group>
                         <group name="action_details">
                             <field name="usage"/>
@@ -183,9 +181,7 @@
                                     <field name="domain"/>
                                     <field name="context"/>
                                     <field name="limit"/>
-                                    <field name="auto_search"/>
                                     <field name="filter"/>
-                                    <field name="multi"/>
                                 </group>
                             </group>
                             <group name="help" string="Help">

--- a/odoo/addons/base/views/ir_ui_view_views.xml
+++ b/odoo/addons/base/views/ir_ui_view_views.xml
@@ -121,7 +121,7 @@
         </record>
         <act_window id="reset_view_arch_wizard_action"
             name="Reset View Architecture"
-            src_model="ir.ui.view"
+            binding_model="ir.ui.view" binding_views="form"
             res_model="reset.view.arch.wizard"
             view_type="form"
             view_mode="form"

--- a/odoo/addons/base/views/res_users_views.xml
+++ b/odoo/addons/base/views/res_users_views.xml
@@ -29,10 +29,9 @@
         </record>
         <act_window id="change_password_wizard_action"
             name="Change Password"
-            src_model="res.users"
+            binding_model="res.users"
             res_model="change.password.wizard"
-            view_type="form" view_mode="form"
-            key2="client_action_multi" target="new"
+            view_type="form" view_mode="form" target="new"
             groups="base.group_erp_manager"/>
 
         <!-- res.groups -->

--- a/odoo/addons/base/wizard/base_partner_merge_views.xml
+++ b/odoo/addons/base/wizard/base_partner_merge_views.xml
@@ -102,10 +102,9 @@
         <act_window
             id="action_partner_merge"
             res_model="base.partner.merge.automatic.wizard"
-            src_model="res.partner"
+            binding_model="res.partner"
+            binding_views="list"
             target="new"
-            multi="True"
-            key2="client_action_multi"
             view_mode="form"
             name="Merge Contacts"/>
 </odoo>
